### PR TITLE
weather-mv: EE ingestion pipeline

### LINKED ISSUE #177: Extend weather-mv to ingest data into Earth Engine (EE)

### PR DESCRIPTION
**Pull request by deepgabani8**
Fixes #177 

The functionality to ingest grib data into the earth engine.

It will convert the grib data into the data format which the earth engine supports at the moment - GeoTiff (.tiff).
It will then upload the tiff image into the earth engine.

### CODE REVIEW

File: weather_mv/README.md
Comment (fredzyda) [inline weather_mv/README.md:19]: Engine.

File: weather_mv/loader_pipeline/ee.py
Comment (fredzyda) [inline weather_mv/loader_pipeline/ee.py:259]: I think this is the wrong driver. I think you should be using the COG driver. Documentation for it is here: https://gdal.org/drivers/raster/cog.html#raster-cog

File: weather_mv/loader_pipeline/ee.py
Comment (fredzyda) [inline weather_mv/loader_pipeline/ee.py:272]: I'm a little confused by this part. Can't you just have rasterio open up an in-memory buffer and then pass that buffer directly to the shutil.copyfileobj call? This seems like a disk write that you don't have to do...

File: weather_mv/loader_pipeline/ee.py
Comment (fredzyda) [inline weather_mv/loader_pipeline/ee.py:308]: Oh, you're doing ingestion here? I'm not sure this is what we wanted. If you write out a COG, you can then write a COG-backed asset directly to Earth Engine and avoid having to wait for an ingestion task to finish. This is way faster and generally what we use. We should have a conversation about wether we want COG-backed ingestion or geotiff to native ingestion here. Perhaps both, but I think COG-backed assets are what we'll want to use most of the time.

File: weather_mv/loader_pipeline/ee.py
Comment (fredzyda) [inline weather_mv/loader_pipeline/ee.py:305]: For a COG-backed asset, the call you want here is ee.data.createAsset. That will return immediately and not start a task. See the documentation here: https://developers.google.com/earth-engine/apidocs/ee-data-createasset

**Review by fredzyda — CHANGES_REQUESTED** [review]
I think this is a fine implementation of EE ingestion to internal EE assets, but I think the COG backed assets are more useful for this tool given how much faster the ingestion is. We should talk about which ingestion technique we want to support or if we should just go ahead and support both with a flag. I will also mention this to @alxmrs to see what his opinion is.

File: weather_mv/loader_pipeline/util.py
Comment (alxmrs) [inline weather_mv/loader_pipeline/util.py:41]: Thanks for moving this to utils.

File: weather_mv/loader_pipeline/ee.py
Comment (deepgabani8) [reply weather_mv/loader_pipeline/ee.py]: Agreed. Updated it.

File: weather_mv/loader_pipeline/ee.py
Comment (deepgabani8) [reply weather_mv/loader_pipeline/ee.py:237]: Oh I added the dry run check in ToEarthEngine's expand. It will just log these path uris and won't go inside any transforms. WDYT? Should we let it go inside each transform in case of dry run?

```
def expand(self, paths):
        """Converts gribs from paths into tiff and uploads them into the earth engine."""
        if not self.dry_run:
            (
                paths
                | 'ConvertToCog' >> beam.ParDo(
                    ConvertToCog(
                        tiff_location=self.tiff_location,
                        open_dataset_kwargs=self.xarray_open_dataset_kwargs,
                        disable_in_memory_copy=self.disable_in_memory_copy,
                        disable_grib_schema_normalization=self.disable_grib_schema_normalization))
                | 'IngestIntoEE' >> beam.ParDo(
                    IngestIntoEE(
                        ee_asset=self.ee_asset,
                        use_personal_account=self.use_personal_account))
            )
        else:
            (
                paths
                | 'Log Grib Files' >> beam.Map(logger.debug)
            )
```

File: weather_mv/loader_pipeline/ee.py
Comment (alxmrs) [inline weather_mv/loader_pipeline/ee.py:265]: Nice, this is a useful class.

File: weather_mv/loader_pipeline/util.py
Comment (alxmrs) [inline weather_mv/loader_pipeline/util.py:283]: Excellent RateLimit utility!

**Review by alxmrs — APPROVED** [review]
Added some notes, but no major changes requested. LGTM, Given @fredzyda gives it the 👍🏽 too, this is ready for merge (after requested changes are addressed).

File: weather_mv/loader_pipeline/ee.py
Comment (fredzyda) [reply weather_mv/loader_pipeline/ee.py:437]: I agree, I think the only reason to catch and then re-raise this exception would be to ensure logging happened. Given that the retry is in place, it may be that the first exceptions may not be logged as it retries.

**Review by fredzyda — COMMENTED** [review]
Generally looks good to me. I'm happy to merge this when @alxmrs is satisfied.

**Comment by alxmrs** [comment]
Our CI system was broken due to downstream dependency changes. I've just pushed a fix in #193. Please merge from the main branch to get our tests to work before the final merge.

File: weather_mv/loader_pipeline/ee_test.py
Comment (deepgabani8) [reply weather_mv/loader_pipeline/ee_test.py:67]: Yes, thank you.

File: weather_mv/loader_pipeline/ee_test.py
Comment (deepgabani8) [reply weather_mv/loader_pipeline/ee_test.py:74]: `Caveat: I don't like how the comment the AAA section names, that's unnecessary` would you elaborate this please?

File: weather_mv/loader_pipeline/ee.py
Comment (deepgabani8) [reply weather_mv/loader_pipeline/ee.py:437]: Yes, logging the error.

File: weather_mv/loader_pipeline/ee.py
Comment (deepgabani8) [reply weather_mv/loader_pipeline/ee.py]: Right, created the issue #197.

File: weather_mv/loader_pipeline/ee_test.py
Comment (alxmrs) [reply weather_mv/loader_pipeline/ee_test.py]: Sure thing: I think that including comments like `# Arrange`, `# Act`, and `# Assert` are unnecessary. However, using whitespace to distinguish between each section is great.

File: weather_mv/loader_pipeline/sinks.py
Comment (alxmrs) [reply weather_mv/loader_pipeline/sinks.py:121]: Ah, thanks for checking. Your original formatting LGTM in that case.

**Review by fredzyda — APPROVED** [review]

**Pull request merged**
